pkg/service/dbaas/service.go

```diff
@@ -0,0 +1,229 @@
+package dbaas
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pipeline "github.com/ccremer/go-command-pipeline"
+	"github.com/vshn/exoscale-metrics-collector/pkg/clients/exoscale"
+	"github.com/vshn/exoscale-metrics-collector/pkg/database"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	egoscale "github.com/exoscale/egoscale/v2"
+	db "github.com/vshn/exoscale-metrics-collector/pkg/database"
+	"github.com/vshn/exoscale-metrics-collector/pkg/service"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	groupVersionResources = map[string]schema.GroupVersionResource{
+		"pg": {
+			Group:    "exoscale.crossplane.io",
+			Version:  "v1",
+			Resource: "postgresqls",
+		},
+	}
+)
+
+// Detail a helper structure for intermediate operations
+type Detail struct {
+	Organization, DBName, Namespace, Plan, Zone string
+}
+
+// Context is the context of the DBaaS service
+type Context struct {
+	context.Context
+	dbaasDetails     []Detail
+	exoscaleDBaasS   []*egoscale.DatabaseService
+	aggregatedDBaasS map[db.Key]db.Aggregated
+}
+
+// Service is the context of the DBaaS service
+type Service struct {
+	exoscaleClient *egoscale.Client
+	k8sClient      dynamic.Interface
+	database       *db.DBaaSDatabase
+}
+
+// NewDBaaSService creates a Service with the initial setup
+func NewDBaaSService(exoscaleClient *egoscale.Client, k8sClient dynamic.Interface, databaseURL string) Service {
+	return Service{
+		exoscaleClient: exoscaleClient,
+		k8sClient:      k8sClient,
+		database: &db.DBaaSDatabase{
+			Database: db.Database{
+				URL: databaseURL,
+			},
+		},
+	}
+}
+
+// Execute executes the main business logic for this application by gathering, matching and saving data to the database
+func (s *Service) Execute(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Running metrics collector by step")
+
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.NewStep("Fetch cluster managed DBaaS", s.fetchManagedDBaaS),
+		p.NewStep("Fetch exoscale DBaaS usage", s.fetchDBaaSUsage),
+		p.NewStep("Filter supported DBaaS", filterSupportedServiceUsage),
+		p.NewStep("Aggregate DBaaS services by namespace and plan", aggregateDBaaS),
+		p.WithNestedSteps("Save to billing database", hasAggregatedInstances,
+			p.NewStep("Get billing date", s.getBillingDate),
+			p.NewStep("Save to database", s.saveToDatabase),
+		),
+	)
```

Comment:
I know it's already all over the codebase so we don't (shouldn't) change this now, but I honestly think these Pipelines just make everything harder to read with no clear benefit..

Reply:
I agree, it's not very go-like style, but it's hard to change now.